Hyperopt: Add threading backend wrapper to avoid loky IPC crashes

Papy-Boyington opened the pull request:
Summary
Force the Hyperopt parallel execution to use the threading backend to avoid loky IPC/semaphore crashes under heavy load.

Quick changelog
- Hyperopt: wrap joblib.Parallel(...) inside parallel_config(backend="threading", inner_max_num_threads=1).
- Stability: avoids SemLock._rebuild FileNotFoundError / TerminatedWorkerError observed with loky at ≥2 workers.
- Performance hygiene: inner_max_num_threads=1 prevents BLAS/OMP oversubscription per job.

What's new?
When Hyperopt runs with loky (multiprocessing) and multiple workers, some environments hit intermittent IPC/semaphore issues (POSIX semaphores in /dev/shm), leading to:
- SemLock._rebuild FileNotFoundError during worker startup, and TerminatedWorkerError reported by joblib.
- This PR forces the threading backend around the existing Parallel(...) call: from joblib import Parallel, delayed, parallel_config

with parallel_config(backend="threading", inner_max_num_threads=1):
    results = Parallel(n_jobs=self.n_jobs)(
        delayed(optimizer_wrapper)(v) for v in asked
    )

Impact:
- No API or CLI changes.
- Stability improves for heavy Hyperopt runs (e.g., OnlyProfit).
- Performance remains good when workloads release the GIL (NumPy/Pandas kernels).
- If desired in a follow-up, this could be made configurable (e.g., joblib_backend: "loky"|"threading"), keeping loky as default.

Tests & style:
- All existing tests pass locally (pytest).
- Code is PEP8-compliant; pre-commit/ruff passes locally.
- Happy to add a small Hyperopt integration test using the threading backend if requested.

AI usage disclosure:
Changes were authored with assistance from GPT-5 Thinking (AI). I reviewed and validated the patch and its rationale.

<!-- Thank you for sending your pull request. But first, have you included
unit tests, and is your code PEP8 conformant? [More details](https://github.com/freqtrade/freqtrade/blob/develop/CONTRIBUTING.md)

Did you use AI to create your changes?
If so, please state it clearly in the PR description (failing to do so may result in your PR being closed).

Also, please do a self review of the changes made before submitting the PR to make sure only relevant changes are included.
-->
## Summary

<!-- Explain in one sentence the goal of this PR -->

Solve the issue: #___

## Quick changelog

- <change log 1>
- <change log 1>

## What's new?

<!-- Explain in details what this PR solve or improve. You can include visuals. -->

Papy-Boyington committed:
Hyperopt: Add threading backend wrapper to avoid loky IPC crashes

Summary
Force the Hyperopt parallel execution to use the threading backend to avoid loky IPC/semaphore crashes under heavy load.

Quick changelog
- Hyperopt: wrap joblib.Parallel(...) inside parallel_config(backend="threading", inner_max_num_threads=1).
- Stability: avoids SemLock._rebuild FileNotFoundError / TerminatedWorkerError observed with loky at ≥2 workers.
- Performance hygiene: inner_max_num_threads=1 prevents BLAS/OMP oversubscription per job.

What's new?
When Hyperopt runs with loky (multiprocessing) and multiple workers, some environments hit intermittent IPC/semaphore issues (POSIX semaphores in /dev/shm), leading to:
- SemLock._rebuild FileNotFoundError during worker startup, and TerminatedWorkerError reported by joblib.
- This PR forces the threading backend around the existing Parallel(...) call:
from joblib import Parallel, delayed, parallel_config

with parallel_config(backend="threading", inner_max_num_threads=1):
    results = Parallel(n_jobs=self.n_jobs)(
        delayed(optimizer_wrapper)(v) for v in asked
    )

Impact:
- No API or CLI changes.
- Stability improves for heavy Hyperopt runs (e.g., OnlyProfit).
- Performance remains good when workloads release the GIL (NumPy/Pandas kernels).
- If desired in a follow-up, this could be made configurable (e.g., joblib_backend: "loky"|"threading"), keeping loky as default.

Tests & style:
- All existing tests pass locally (pytest).
- Code is PEP8-compliant; pre-commit/ruff passes locally.
- Happy to add a small Hyperopt integration test using the threading backend if requested.

AI usage disclosure:
Changes were authored with assistance from GPT-5 Thinking (AI). I reviewed and validated the patch and its rationale.

xmatthias commented on freqtrade/optimize/hyperopt/hyperopt.py:
```diff
@@ -14,7 +14,13 @@
+    from joblib import parallel_config as _joblib_parallel_ctx
+except ImportError:  # joblib < 1.1 fallback
+    # `parallel_backend` offers a similar context manager interface
+    from joblib import parallel_backend as _joblib_parallel_ctx  # type: ignore
```
Freqtrade pins Joblib to a particular version - and assumes that users follow the installation and update documentation.
based on this is - are you sure this is necessary?

Papy-Boyington replied:
No (and I know), and you're right, it's not necessary.
This is just artificial intelligence code. And I left it in just in case...
I was pretty sure you wouldn't find it relevant ;)
This piece of code is pretty ugly...

xmatthias commented on freqtrade/optimize/hyperopt/hyperopt.py:
```diff
@@ -147,17 +153,24 @@ def print_results(self, results: dict[str, Any]) -> None:
+        n_jobs = getattr(parallel, "n_jobs", self.n_jobs)
+
+        # Use THREADING backend (no POSIX semaphores/IPC) and avoid nested BLAS/OMP threads per job
+        with _joblib_parallel_ctx(backend="threading", inner_max_num_threads=1):
```
I don't think using a "threading" backend here is sensible or reasonable.

Currently, backtesting runs on for example 4 cores in parallel  (assuming you have 4 cores and use `-j 4` - otherwise it defaults to "available -1")- and only once all 4 batched backtests finish, it'll fetch new values to optimize.

with threading - the 4 backtests will now run serially (due to the GIL), greatly reducing the overall speed of hyperopt as the overall parallelism is lost.

Can you point to any documentation that points to the fact that joblib will not do that when using "threading" as backend??
If not and if joblib is acting in threading mode the way i'd expect (which is - use multiple threads, not multiple processes) - then please consider this as "this is a bad idea, we'll reject this" comment.

Papy-Boyington replied:
You raise a crucial point about the GIL.
The joblib documentation itself suggests the threading backend is viable for this specific use case, as much of the workload is in C extensions that release the GIL.
Here is the link to their documentation:
https://joblib.readthedocs.io/en/latest/parallel.html#thread-based-parallelism-vs-process-based-parallelism
The relevant quote is:
"threading is a good choice for I/O-bound tasks or tasks involving a lot of calls to C extensions that release the GIL."
Since Freqtrade's backtests heavily rely on pandas and TA-Lib, the performance degradation shouldn't be purely serial.
I think people on systems (like macOS and even my Linux :/) where loky fails with the FileNotFoundError, making hyperopt unusable for them.
Let me know what you think.

xmatthias replied:
i think you're wrong and had AI suggest a non-suitable solution.

90% of the time of hyperopt is NOT spent in c extensions - but in the simulation phase.
essentially none of the work is I/O bound.
Also - C extension doesn't automatically mean "releases the GIL".

so the impact here will be about 90% of the time being spent in serial backtesting - which you can accomplish just as well with `-j 1`.

xmatthias replied:
To be quite frank - i don't think this is a PR we're very likely to accept (leaving this open for further discussion for the moment)

I don't see anything it provides aside from downsides - a seeming fix (use `-j 1`, it'll fix it all the same) - but with a speed impact beyond reason - making every line this adds a liability / potential maintenance problem sometime in the future (that's always a consideration we need to take before accepting a PR).


I'd also apreciate _discussion_, rather than _new commits_ to this (which is noise - but won't change facts).
if that continues, i'll assume that it's just your AI continuing, and will close it to avoid further noise about it.

Papy-Boyington committed:
Update cli_options.py

Add this code for correcting and update this commit : https://github.com/freqtrade/freqtrade/pull/12205/commits

    "joblib_backend": Arg(
        "--joblib-backend",
        help="Parallelism backend for Hyperopt: 'loky' (processes, default) or 'threading' (no IPC).",
        choices=("loky", "threading"),
        default=None,
    ),

Papy-Boyington committed:
Add this code for correcting and update this commit : https://github.com/freqtrade/freqtrade/pull/12205/commits

Update def start_hyperopt

Papy-Boyington committed:
Add this code for correcting and update this commit : https://github.com/freqtrade/freqtrade/pull/12205/commits

Add     "joblib_backend", in ARGS_HYPEROPT

Papy-Boyington committed:
Update hyperopt.py and Add this code for correcting and update this commit : https://github.com/freqtrade/freqtrade/pull/12205/commits

Update def run_optimizer_parallel AFTER succefull tests

Papy-Boyington commented:
Hi,

First of all thank you for this amazing project ! :)

I have tryed to fix this: I ran the test with two processors and 128 GB of RAM. (On Debian 12 Python 3.11.) > Crash. And Only  OnlyProfitHyperOptLoss crashed consistently, while SharpeHyperOptLoss was fine?!?

2025-08-29 18:31:25,397 - freqtrade.optimize.hyperopt.hyperopt - INFO - Found 32 CPU cores. Let's make them scream!
2025-08-29 18:31:25,398 - freqtrade.optimize.hyperopt.hyperopt - INFO - Number of parallel jobs set as: 16
2025-08-29 18:31:25,400 - freqtrade.optimize.hyperopt.hyperopt_optimizer - INFO - Using optuna sampler NSGAIIISampler.
2025-08-29 18:31:25,450 - freqtrade.optimize.hyperopt.hyperopt - INFO - Effective number of parallel workers used: 16


--------------------------------------------------------------------------------
LokyProcess-117 failed with traceback: 
--------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/iadmin/freqtrade/.venv/lib/python3.11/site-packages/joblib/externals/loky/backend/popen_loky_posix.py", line 176, in <module>
    process_obj = pickle.load(from_parent)
                  ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/iadmin/freqtrade/.venv/lib/python3.11/site-packages/joblib/externals/loky/backend/synchronize.py", line 131, in __setstate__
    self._semlock = _SemLock._rebuild(*state)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory


--------------------------------------------------------------------------------


--------------------------------------------------------------------------------
LokyProcess-118 failed with traceback: 
--------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/iadmin/freqtrade/.venv/lib/python3.11/site-packages/joblib/externals/loky/backend/popen_loky_posix.py", line 176, in <module>
    process_obj = pickle.load(from_parent)
                  ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/iadmin/freqtrade/.venv/lib/python3.11/site-packages/joblib/externals/loky/backend/synchronize.py", line 131, in __setstate__
    self._semlock = _SemLock._rebuild(*state)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory


--------------------------------------------------------------------------------
                                                                                            Hyperopt results                                                                                            
                      ┏━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━┓                      
                      ┃ Best   ┃    Epoch ┃ Trades ┃    Win  Draw  Loss  Win% ┃ Avg profit ┃                     Profit ┃ Avg duration ┃     Objective ┃     Max Drawdown (Acct) ┃                      
                      ┡━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━┩                      
                      │ * Best │   7/1200 │   6471 │   2812   457  3202  43.5 │     -0.02% │    -81.643 USDT  (-11.66%) │      2:01:00 │      81.64284 │ 110.601 USDT   (15.27%) │                      
                      │ * Best │  12/1200 │   1088 │    378    39   671  34.7 │     -0.07% │    -66.110 USDT   (-9.44%) │      1:10:00 │      66.10982 │  76.511 USDT   (10.77%) │                      
                      │ * Best │  28/1200 │  24467 │  9513   359  14595  38.9 │      0.03% │   1150.155 USDT  (164.31%) │      0:39:00 │  -1,150.15497 │ 110.238 USDT    (5.76%) │                      
                      │ * Best │  29/1200 │  24300 │ 12005     1  12294  49.4 │      0.07% │   4427.047 USDT  (632.44%) │      0:51:00 │  -4,427.04723 │ 133.820 USDT    (2.70%) │                      
                      │ Best   │  74/1200 │  21857 │  10667  1629  9561  48.8 │      0.08% │   4833.612 USDT  (690.52%) │      1:13:00 │  -4,833.61242 │ 145.832 USDT    (2.84%) │                      
                      │ Best   │ 136/1200 │  22425 │ 11142     7  11276  49.7 │      0.10% │  9280.754 USDT (1,325.82%) │      1:09:00 │  -9,280.75420 │ 238.097 USDT    (2.59%) │                      
                      │ Best   │ 147/1200 │  21340 │ 10752    14  10574  50.4 │      0.12% │ 14705.387 USDT (2,100.77%) │      1:19:00 │ -14,705.38664 │ 283.195 USDT    (2.14%) │                      
                      │ Best   │ 166/1200 │  21375 │ 10778    14  10583  50.4 │      0.12% │ 15234.323 USDT (2,176.33%) │      1:20:00 │ -15,234.32334 │ 285.274 USDT    (2.09%) │                      
                      │ Best   │ 171/1200 │  21048 │ 10677   330  10041  50.7 │      0.12% │ 15528.292 USDT (2,218.33%) │      1:23:00 │ -15,528.29177 │ 310.576 USDT    (2.28%) │                      
                      │ Best   │ 174/1200 │  20919 │  10586   338  9995  50.6 │      0.13% │ 16436.705 USDT (2,348.10%) │      1:25:00 │ -16,436.70499 │ 332.875 USDT    (2.29%) │                      
                      │ Best   │ 192/1200 │  20886 │ 10611   260  10015  50.8 │      0.13% │ 16625.266 USDT (2,375.04%) │      1:26:00 │ -16,625.26611 │ 352.916 USDT    (2.44%) │                      
                      │ Best   │ 202/1200 │  21555 │ 10694   204  10657  49.6 │      0.13% │ 21259.001 USDT (3,037.00%) │      1:19:00 │ -21,259.00132 │ 405.861 USDT    (2.26%) │                      
                      └────────┴──────────┴────────┴──────────────────────────┴────────────┴────────────────────────────┴──────────────┴───────────────┴─────────────────────────┘                      
Epochs ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━                                                                                                                                   224/1200  19% • 1:55:50 • 0:10:56
2025-08-29 20:27:15,931 - freqtrade - ERROR - Fatal exception!
Traceback (most recent call last):
  File "/home/iadmin/freqtrade/.venv/lib/python3.11/site-packages/freqtrade/main.py", line 47, in main
    return_code = args["func"](args)
                  ^^^^^^^^^^^^^^^^^^
  File "/home/iadmin/freqtrade/.venv/lib/python3.11/site-packages/freqtrade/commands/optimize_commands.py", line 111, in start_hyperopt
    hyperopt.start()
  File "/home/iadmin/freqtrade/.venv/lib/python3.11/site-packages/freqtrade/optimize/hyperopt/hyperopt.py", line 295, in start
    f_val = self.run_optimizer_parallel(
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/iadmin/freqtrade/.venv/lib/python3.11/site-packages/freqtrade/optimize/hyperopt/hyperopt.py", line 160, in run_optimizer_parallel
    return parallel(optimizer_wrapper(v) for v in asked)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/iadmin/freqtrade/.venv/lib/python3.11/site-packages/joblib/parallel.py", line 2072, in __call__
    return output if self.return_generator else list(output)
                                                ^^^^^^^^^^^^
  File "/home/iadmin/freqtrade/.venv/lib/python3.11/site-packages/joblib/parallel.py", line 1682, in _get_outputs
    yield from self._retrieve()
  File "/home/iadmin/freqtrade/.venv/lib/python3.11/site-packages/joblib/parallel.py", line 1784, in _retrieve
    self._raise_error_fast()
  File "/home/iadmin/freqtrade/.venv/lib/python3.11/site-packages/joblib/parallel.py", line 1859, in _raise_error_fast
    error_job.get_result(self.timeout)
  File "/home/iadmin/freqtrade/.venv/lib/python3.11/site-packages/joblib/parallel.py", line 758, in get_result
    return self._return_or_raise()
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/iadmin/freqtrade/.venv/lib/python3.11/site-packages/joblib/parallel.py", line 773, in _return_or_raise
    raise self._result
joblib.externals.loky.process_executor.TerminatedWorkerError: A worker process managed by the executor was unexpectedly terminated. This could be caused by a segmentation fault while calling the 
function or by an excessive memory usage causing the Operating System to kill the worker.

xmatthias commented:
> I have tryed to fix this: I ran the test with two processors and 128 GB of RAM. (On Debian 12 Python 3.11.) > Crash. And Only OnlyProfitHyperOptLoss crashed consistently, while SharpeHyperOptLoss was fine?!?

This pretty much points to the fact that the error messsage below is actually what's happening - your process simply runs out of memory.

> [...] or by an excessive memory usage causing the Operating System to kill the worker.

OnlyProfitHyperOptLoss is very likely to produce more trades - while SharpeHyperOptLoss is a bit more sophisticated, optimizing also on other things (which usually means - fewer trades which in term means - fewer resources consumed.

Switching to threading will solve this - because you've now moved your whole process to run on a single core (observe the CPU load with htop or similar - you'll see that your PR will no longer use all 4 cores.

If that's what you're after - then `-j 1` can accomplish the same thing - run just one job after the other - which is what threading will resolve to.

---

Unfortunately - the code looks pretty much like raw AI code - which i'm pretty certain it is - with the expected idiocy it has added - like fallbacks for scenarios freqtrade doesn't support and similar.

Before even looking at it - please do a **good** manual review yourself and remove all pointless and unnecessary changes (the import section is a good starting point to start on removing stuff - but it's far from the only place where absolute unnecessary stuff happens).

Papy-Boyington committed:
Remove ugly AI code in hyperopt.py

Papy-Boyington committed:
Update hyperopt.py in Shakespeare language and clean code

Papy-Boyington committed:
Update hyperopt.py

Papy-Boyington committed:
Update hyperopt.py actually testing it with 22 processors

Papy-Boyington commented:
I personally checked the code. It seems there's no way to solve my problem.
For the same strategy, 22 processors are OK with SharpeHyperOptLoss.
While 2 processors are NOK with OnlyProfitHyperOptLoss. Only with one processor is OK…
I give up. After 3 weeks of intensive research...

stash86 commented:
why onlyprofit crashed is because as the name says, it favor profit only. So the optimizer with favor a big number of trades with very low avg profit, which in turns will kill your RAM very fast. Why? Because we store every trade results of every epoch, so if each epoch has gigantic trades, the RAM usage also exponentially high.

This is not FT issue, but more strategy issue.

xmatthias commented:
Well that's what i said all along
Independent of the timerange - your last (shown) epoch does 21555 trades.
that's a list of 21555 entries - with a sublist with _at least_ 2 entries for orders (maybe more - depends on the strategy).

for that whole thing - freqtrade needs to build statistics, analyze it, ... 
switching to threading works because the process becomes much more serial - and yes - eliminates multiprocessing (but at the expense of performance) so you can probably get to the same result with `-j 1` within a similar time.


---

```
Traceback (most recent call last):
File "/home/iadmin/freqtrade/.venv/lib/python3.11/site-packages/joblib/externals/loky/backend/popen_loky_posix.py", line 176, in
process_obj = pickle.load(from_parent)
^^^^^^^^^^^^^^^^^^^^^^^^
File "/home/iadmin/freqtrade/.venv/lib/python3.11/site-packages/joblib/externals/loky/backend/synchronize.py", line 131, in setstate
self._semlock = _SemLock._rebuild(*state)
^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory
```

Now i also think that your approach to this is wrong (that's what happens when you ask AI).
It's a joblib exception (triggered by the loky backend, sure) - so the better approach is to look there for the actual, underlying problem.

There's a couple of reports of (almost exactly the same) error in combination with multiprocessing (https://github.com/python/cpython/issues/94765, or even https://github.com/joblib/joblib/issues/1247 or https://github.com/joblib/joblib/issues/1205 which would point to an older joblib version) - notably none  with a python version > 3.12.2 (or even 3.13) - suggesting that (while the issue is still open) it's quite possible that it's fixed in a more recent python version.
that's also the problem of having a PR rather than an issue - if you're on 3.11.1 or some other low minor version, we would actually know that it has a couple of problems - not just in this area.
The same is true for the joblib version - we assume that the pinned version is used - but we neither enforce nor check that actively - so that might also be worth checking on your side.

xmatthias commented:
with loky backend: 

<img width="772" height="287" alt="image" src="https://github.com/user-attachments/assets/c04b8971-cfb3-432d-8d16-7a9ac05c71fd" />

Same run - with threading:

```
reqtrade.strategy.strategy_wrapper - ERROR - Unexpected error dictionary changed size during iteration calling 
<bound method TurtleStrategyX5.custom_roi of <TurtleStrategyX5.TurtleStrategyX5 object at 0x7fc1f309b4d0>>
Traceback (most recent call last):
  File "/home/xmatt/devel/cryptos/freqtrade/freqtrade/strategy/strategy_wrapper.py", line 30, in wrapper
    kwargs["trade"] = deepcopy(kwargs["trade"])
                      ~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/home/xmatt/.local/share/uv/python/cpython-3.13.7-linux-x86_64-gnu/lib/python3.13/copy.py", line 163, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/xmatt/.local/share/uv/python/cpython-3.13.7-linux-x86_64-gnu/lib/python3.13/copy.py", line 260, in _reconstruct
    state = deepcopy(state, memo)
  File "/home/xmatt/.local/share/uv/python/cpython-3.13.7-linux-x86_64-gnu/lib/python3.13/copy.py", line 137, in deepcopy
    y = copier(x, memo)
  File "/home/xmatt/.local/share/uv/python/cpython-3.13.7-linux-x86_64-gnu/lib/python3.13/copy.py", line 221, in _deepcopy_dict
    for key, value in x.items():
```

Even if this would be working (it isn't functional at the moment) - threading backend wouldn't be able to accomplish the parallelism hyperopt needs - and would run one process after the ohter.

As this is non-functional - and from a concept perspective wrong (a missguided AI - same result can be accomplished with `-j1`) - i'm going to close this - as we have no plans in accepting this or working on this further.
thanks for your understanding.
